File: IHM/test1-ROSWebTools/ros2-nodes/loki_node.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped
from std_msgs.msg import UInt64
from rclpy.time import Time
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy

# Import your custom message
from interfaces.msg import CommandTimestamp  # Replace 'your_package_name' with the actual package name

class LokiNode(Node):
    def __init__(self):
        super().__init__('loki_node')

        # Define QoS with best-effort reliability and keep last 10 messages in the queue
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10
        )

        # Subscriptions with optimized QoS
        self.create_subscription(TwistStamped, '/cmd_vel', self.listener_callback, qos_profile)
        self.create_subscription(UInt64, '/temp_completion', self.completion_callback, qos_profile)

        # Publisher for custom message
        self.time_stamp_publisher = self.create_publisher(CommandTimestamp, '/cmd_vel_time_stamp', qos_profile)

    def listener_callback(self, msg):
        # Get current time and convert to nanoseconds
        T1 = self.get_current_time_in_ns()

        # Extract command_id from header's stamp (time in nanoseconds)
        command_id = self.extract_ns_from_header(msg.header.stamp)

        # Publish timestamp for T1 using the custom message
        self.publish_timestamp(command_id, 1, T1)

    def completion_callback(self, msg):
        # Get current time in nanoseconds for T5
        T5 = self.get_current_time_in_ns()
        self.publish_timestamp(msg.data, 5, T5)

    def publish_timestamp(self, command_id, timestamp_index, timestamp_value):
        msg = CommandTimestamp()  # Create an instance of your custom message
        msg.command_id = command_id
        msg.timestamp_index = timestamp_index
        msg.timestamp = timestamp_value
        self.time_stamp_publisher.publish(msg)

    def get_current_time_in_ns(self):
        return self.get_clock().now().nanoseconds
    
    def extract_ns_from_header(self, stamp):
        return Time.from_msg(stamp).nanoseconds

def main(args=None):
    rclpy.init(args=args)
    node = LokiNode()

    # Use non-blocking spin to improve efficiency
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: IHM/test1-ROSWebTools/ros2-nodes/loki_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <std_msgs/msg/u_int64.hpp>
#include "interfaces/msg/command_timestamp.hpp" // Replace with your actual package name

using namespace std::chrono_literals;

class LokiNode : public rclcpp::Node {
public:
    LokiNode()
        : Node("loki_node") {
        
        // Define QoS with best-effort reliability and keep last 10 messages in the queue
        auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10)).best_effort();

        // Subscriptions with optimized QoS
        cmd_vel_subscription_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
            "/cmd_vel",
            qos_profile,
            std::bind(&LokiNode::listener_callback, this, std::placeholders::_1)
        );

        completion_subscription_ = this->create_subscription<std_msgs::msg::UInt64>(
            "/temp_completion",
            qos_profile,
            std::bind(&LokiNode::completion_callback, this, std::placeholders::_1)
        );

        // Publisher for custom message
        time_stamp_publisher_ = this->create_publisher<interfaces::msg::CommandTimestamp>(
            "/cmd_vel_time_stamp",
            qos_profile
        );
    }

private:
    void listener_callback(const geometry_msgs::msg::TwistStamped::SharedPtr msg) {
        // Get current time in nanoseconds
        int64_t T1 = get_current_time_in_ns();

        // Extract command_id from header's stamp (time in nanoseconds)
        uint64_t command_id = extract_ns_from_header(msg->header.stamp);

        // Publish timestamp for T1 using the custom message
        publish_timestamp(command_id, 1, T1);
    }

    void completion_callback(const std_msgs::msg::UInt64::SharedPtr msg) {
        // Get current time in nanoseconds for T5
        int64_t T5 = get_current_time_in_ns();
        publish_timestamp(msg->data, 5, T5);
    }

    void publish_timestamp(uint64_t command_id, int timestamp_index, int64_t timestamp_value) {
        auto msg = interfaces::msg::CommandTimestamp(); // Create an instance of your custom message
        msg.command_id = command_id;
        msg.timestamp_index = timestamp_index;
        msg.timestamp = timestamp_value;
        time_stamp_publisher_->publish(msg);
    }

    int64_t get_current_time_in_ns() {
        return this->now().nanoseconds();
    }

    uint64_t extract_ns_from_header(const rclcpp::Time& stamp) {
        return stamp.nanoseconds();
    }

    rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr cmd_vel_subscription_;
    rclcpp::Subscription<std_msgs::msg::UInt64>::SharedPtr completion_subscription_;
    rclcpp::Publisher<interfaces::msg::CommandTimestamp>::SharedPtr time_stamp_publisher_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LokiNode>());
    rclcpp::shutdown();
    return 0;
}
